src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -547,21 +595,233 @@ int SketchObject::solve(bool updateGeoAfterSolving /*=true*/)
+        // this->Constraints.touch();
     }
 
+    signalSolverUpdate();
+
     return err;
 }
 
+namespace bg = boost::geometry;
+namespace bgi = boost::geometry::index;
+
+BOOST_GEOMETRY_REGISTER_POINT_3D(
+        Base::Vector3d,double,bg::cs::cartesian,x,y,z)
+
+class SketchObject::GeoHistory {
+public:
+    typedef bgi::linear<16> Parameters;
+
+    typedef std::set<long> IdSet;
+    typedef std::pair<IdSet,IdSet> IdSets;
+    typedef std::list<IdSet> AdjList;
+
+    //associate a geo with connected ones on both points
+    typedef std::map<long, IdSets> AdjMap;
+
+    // maps start/end points to all existing geo to query and update adjacencies
+    typedef std::pair<Base::Vector3d,AdjList::iterator> Value;
+
+    AdjList adjlist;
+    AdjMap adjmap;
+    bgi::rtree<Value,Parameters> rtree;
+
+    AdjList::iterator find(const Base::Vector3d &pt,bool strict=true){
+        std::vector<Value> ret;
+        rtree.query(bgi::nearest(pt,1),std::back_inserter(ret));
+        if(ret.size()) {
+            // NOTE: we are using square distance here, the 1e-6 threshold is
+            // very forgiving. We should have used Precision::SquareConfisuion(),
+            // which is 1e-14. However, there is a problem with current
+            // commandGeoCreate. They create new geometry with initial point of
+            // the exact mouse position, instead of the pre-selected point
+            // position, and rely on auto constraint to snap in the new
+            // geometry. So, we cannot use a very strict threshold here.
+            double tol = strict?Precision::SquareConfusion()*10:1e-6;
+            double d = Base::DistanceP2(ret[0].first,pt);
+            if(d<tol) {
+                if(!strict) FC_TRACE("hit " << FC_xyz(pt));
+                return ret[0].second;
+            }
+            if(!strict)
+                FC_TRACE("miss " << FC_xyz(pt) << ", " << FC_xyz(ret[0].first) << ", " << d);
+        }else if(!strict)
+            FC_TRACE("miss " << FC_xyz(pt));
+        return adjlist.end();
+    }
+
+    void clear() {
+        rtree.clear();
+        adjlist.clear();
+    }
+
+    void update(const Base::Vector3d &pt, long id) {
+        FC_TRACE("update " << id << ", " << FC_xyz(pt));
+        auto it = find(pt);
+        if(it==adjlist.end()) {
+            adjlist.emplace_back();
+            it = adjlist.end();
+            --it;
+            rtree.insert(std::make_pair(pt,it));
+        }
+        it->insert(id);
+    }
+
+    void finishUpdate(const std::map<long,int> &geomap) {
+        IdSet oldset;
+        for(auto &idset : adjlist) {
+            oldset.clear();
+            for(long _id : idset) {
+                long id = abs(_id);
+                auto &v = adjmap[id];
+                auto &adj = _id>0?v.first:v.second;
+                for(auto it=adj.begin(),itNext=it;it!=adj.end();it=itNext) {
+                    ++itNext;

```

Comment:
I couldn't find a way to avoid the loop because of the IDs stored to be swapped back later plus the else if.  So I rewrote slightly with comments to make intent clearer.